distributed/scheduler.py

```diff
@@ -5648,143 +5650,172 @@ async def gather(self, keys, serializers=None):
         return result
 
     @log_errors
-    async def restart(self, client=None, timeout=30, wait_for_workers=True):
+    async def restart(self, client=None, timeout=30):
         """
-        Restart all workers. Reset local state. Optionally wait for workers to return.
+        Reset local state and restart all workers
 
-        Workers without nannies are shut down, hoping an external deployment system
-        will restart them. Therefore, if not using nannies and your deployment system
-        does not automatically restart workers, ``restart`` will just shut down all
-        workers, then time out!
+        This methods expects all workers to have nannies to be able to restart them.
+        If workers without nannies exist, ``restart`` will return a serialized
+        ``ValueError`` that lists the workers without nannies. Consider removing
+        those workers and calling ``restart`` again afterward.
 
-        After ``restart``, all connected workers are new, regardless of whether ``TimeoutError``
-        was raised. Any workers that failed to shut down in time are removed, and
-        may or may not shut down on their own in the future.
+        If ``restart`` fails or times out, it will return a serialized error and
+        leave the cluster in an undefined state.
 
         Parameters
         ----------
         timeout:
-            How long to wait for workers to shut down and come back, if ``wait_for_workers``
-            is True, otherwise just how long to wait for workers to shut down.
-            Raises ``asyncio.TimeoutError`` if this is exceeded.
-        wait_for_workers:
-            Whether to wait for all workers to reconnect, or just for them to shut down
-            (default True). Use ``restart(wait_for_workers=False)`` combined with
-            :meth:`Client.wait_for_workers` for granular control over how many workers to
-            wait for.
+            ``restart`` fails and returns an error if it takes more than ``timeout``
+            seconds
 
-        See also
+        See Also
         --------
         Client.restart
-        Client.restart_workers
+        Scheduler.restart_workers
         """
+        try:
+            await self._restart(client, timeout)
+            return {"status": "OK"}
+        except Exception as e:
+            logger.exception(e)
+            return error_message(e)
```

Comment:
Why manually return the serialized error? This seems like an unusual pattern. I'd rather just raise an error in the function and have the RPC handle serializing and deserializing it if possible?

If it's just about the error message being hard to understand for users, that's a different topic: https://github.com/dask/distributed/issues/4880.

Reply:
Chalk that up to me not knowing any better. You're suggesting

```suggestion
            raise
```
?